components/brave_wallet/renderer/js_ethereum_provider.cc

```diff
@@ -643,4 +649,128 @@ void JSEthereumProvider::MessageEvent(const std::string& subscription_id,
   FireEvent(ethereum::kMessageEvent, event_args);
 }
 
+void JSEthereumProvider::OnProviderRequested() {
+  AnnounceProvider();
+}
+
+void JSEthereumProvider::BindRequestProviderListener() {
+  v8::Isolate* isolate = blink::MainThreadIsolate();
+  v8::HandleScope handle_scope(isolate);
+  v8::Local<v8::Context> context =
+      render_frame()->GetWebFrame()->MainWorldScriptContext();
+
+  auto onProviderRequested = gin::CreateFunctionTemplate(
+      isolate, base::BindRepeating(&JSEthereumProvider::OnProviderRequested,
+                                   weak_ptr_factory_.GetWeakPtr()));
+  auto functionInstance =
+      onProviderRequested->GetFunction(context).ToLocalChecked();
+
+  std::vector<v8::Local<v8::Value>> args;
+  args.push_back(content::V8ValueConverter::Create()->ToV8Value(
+      base::Value("eip6963:requestProvider"), context));
+  args.push_back(std::move(functionInstance));
+
+  CallMethodOfObject(render_frame()->GetWebFrame(), "window",
+                     "addEventListener", std::move(args));
+}
+
+void JSEthereumProvider::AnnounceProvider() {
+  v8::Isolate* isolate = blink::MainThreadIsolate();
+  v8::HandleScope handle_scope(isolate);
+  v8::Local<v8::Context> context =
+      render_frame()->GetWebFrame()->MainWorldScriptContext();
+
+  base::Value::Dict provider_info_value;
+  provider_info_value.Set("rDNS", "com.brave.wallet");
+  provider_info_value.Set("uuid", uuid_);
+  provider_info_value.Set("name", "Brave Wallet");
+  provider_info_value.Set("icon", GetBraveWalletImage());
+
+  auto detail = v8::Object::New(isolate);
+  auto info_object = content::V8ValueConverter::Create()->ToV8Value(
+      std::move(provider_info_value), context);
+
+  if (!info_object.As<v8::Object>()
+           ->SetIntegrityLevel(isolate->GetCurrentContext(),
+                               v8::IntegrityLevel::kFrozen)
+           .ToChecked()) {
+    NOTREACHED();
+    return;
+  }
+
+  if (detail
+          ->Set(context,
+                content::V8ValueConverter::Create()->ToV8Value(
+                    base::Value("info"), context),
+                std::move(info_object))
+          .IsNothing()) {
+    NOTREACHED();
+    return;
+  }
+
+  auto provider =
+      GetProperty(context, context->Global(), kEthereum).ToLocalChecked();
```

Comment:
Does this just grab what's already injected on `window.ethereum`? In the instance where an extension has already overwritten it would this link to our provider object or theirs?

Reply:
At the current moment we rewrite window.ethereum and make it readonly with our provider

Reply:
@kdenhartog this will always use our instance because no extensions can inject `window.ethereum` prior to us.
But that object can definitely be overwritten later by extension so I think we should consider doing `window.braveEthereum` as the main object to deal with eip6963 and `window.ethereum` which still complies to default wallet setting is just an alias to it.
see https://github.com/brave/brave-core/pull/18751#discussion_r1262881024 for more detail